File: pt6/1268.search-suggestions-system.go
package main

import "sort"

func suggestedProducts(products []string, searchWord string) [][]string {
	sort.Strings(products)
	ans := make([][]string, 0, len(searchWord))
	prefix := ""

	for i := 0; i < len(searchWord); i++ {
		prefix += string(searchWord[i])
		start := sort.SearchStrings(products, prefix)
		suggestions := []string{}
		for j := start; j < len(products) && j < start+3; j++ {
			if hasPrefix(products[j], prefix) {
				suggestions = append(suggestions, products[j])
			}
		}
		ans = append(ans, suggestions)
	}

	return ans
}

func hasPrefix(word, prefix string) bool {
	return len(word) >= len(prefix) && word[:len(prefix)] == prefix
}

/*
Explanation

Sort products lexicographically. For each search prefix, binary search for the
first product not less than the prefix, then inspect at most the next three
products and keep those that actually start with the prefix.

Sorting works because all words with a common prefix occupy a contiguous range,
and the first three in that range are the lexicographically smallest
suggestions. This is a compact alternative to a trie.

Edge cases: fewer than three matches; no matches; one product being a prefix of
another.

Time complexity: O(n log n + m log n + 3mL), where m is searchWord length and L
is prefix comparison cost.
Space complexity: O(m) for the returned lists, excluding sorting.
*/

File: pt6/1268.search-suggestions-system.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
    public List<List<String>> suggestedProducts(String[] products, String searchWord) {
        Arrays.sort(products);
        List<List<String>> ans = new ArrayList<>();
        StringBuilder prefix = new StringBuilder();

        for (char ch : searchWord.toCharArray()) {
            prefix.append(ch);
            String current = prefix.toString();
            int start = lowerBound(products, current);
            List<String> suggestions = new ArrayList<>();

            for (int i = start; i < products.length && i < start + 3; i++) {
                if (products[i].startsWith(current)) {
                    suggestions.add(products[i]);
                }
            }
            ans.add(suggestions);
        }

        return ans;
    }

    private int lowerBound(String[] products, String target) {
        int left = 0;
        int right = products.length;
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (products[mid].compareTo(target) < 0) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        return left;
    }
}

/*
Explanation

Sort products lexicographically. For each prefix of searchWord, binary search
for the first product not less than the prefix, then inspect at most the next
three products.

Sorting works because all strings with the same prefix form one contiguous
range. This is a compact alternative to a trie when only three suggestions are
needed.

Edge cases: fewer than three matches; no matches; one product being a prefix of
another.

Time complexity: O(n log n + m log n + 3mL), where m is searchWord length.
Space complexity: O(m) for the answer lists, excluding sorting.
*/
